app/jobs/process_batch_action_job.rb

```diff
@@ -0,0 +1,43 @@
+class ProcessBatchActionJob < ApplicationJob
+  # @param pending_induction_submission_batch [PendingInductionSubmissionBatch]
+  # @param author_email [String]
+  # @param author_name [String]
+  def perform(pending_induction_submission_batch, author_email, author_name)
+    return if pending_induction_submission_batch.processing?
+
+    if pending_induction_submission_batch.processed?
+      # change status first for quick response
+      pending_induction_submission_batch.completed!
+      batch_action(pending_induction_submission_batch, author_email, author_name).do!
```

Comment:
Just flagging that `pass!` calls `record_outcome!` which includes `send_pass_induction_notification_to_trs`.

If this method raises, the internal transaction that `record_outcome!` uses will roll back DB changes inside `record_outcome!`, but **not** things like posting to the TRS API. That means we lose atomicity and if the batch was marked as `completed!` beforehand, we may not catch the inconsistency.

  Would it be safer to: 

  - call `do!` first and only mark the batch as `completed!` if it succeeds?
  - consider compensating or tracking partially applied batches incase of failure?

Reply:
You're right to flag it. The reason is the lag in actioning the changes once the user confirms. This was done so we could render success. There is potentially a need for another status, `actioning` or `doing` 😉 , which comes before `completed`. However, the things it's attempting should all succeed without exception because they have been pre-validated and filtered.

I'll take your comments away and address them in my next CSV related branch. Thanks @avinhurry 🙇🏼‍♂️